optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/Joiners.java

```diff
@@ -147,10 +151,99 @@
      * }
      */
 
-    // TODO
-    // join(..., planningVariableContainsCached(Talk::getPeriod, (Period a, Period b) -> a.overlaps(b)))
-    // get the period value range, does a cartesian product on it, so it maps every period to an overlapping periodList
-    // then keep an index from every period to all talks in an overlapping period (possible the same period)
+    /**
+     * For pairs of intervals [a,b) and [c,d), returns only those pairs where [a,b) and [c,d) overlap (or, in other
+     * words, there exists an x such that x in [a,b) and x in [c,d)).
+     * These are exactly the pairs where {@code a < d} and {@code b > c} (or in other words, when the first interval
+     * starts before the second one ends and the first interval ends after the second interval begins).
+     * 
+     * @param <A> the type of the first argument
+     * @param <B> the type of the second argument
+     * @param <Property_> the type used to define the interval, comparable
+     * 
+     * @param leftStartMapping maps the first argument to the first interval start point (inclusive)
+     * @param leftEndMapping maps the first argument to the first interval end point (exclusive)
+     * @param rightStartMapping maps the second argument to the second interval start point (inclusive)
+     * @param rightEndMapping maps the second argument to the second interval end point (exclusive)
+     * 
+     * @return never null, an indexed joiner that filters the constraint stream to only include elements (A,B) where
+     *         A's and B's intervals (as defined by the function mapping) overlap
+     */
+    public static <A, B, Property_ extends Comparable<Property_>> BiJoiner<A, B> overlaps(
+            Function<A, Property_> leftStartMapping,
+            Function<A, Property_> leftEndMapping,
+            Function<B, Property_> rightStartMapping,
+            Function<B, Property_> rightEndMapping) {
+        return AbstractBiJoiner.merge(Joiners.lessThan(leftStartMapping, rightEndMapping),
+                Joiners.greaterThan(leftEndMapping, rightStartMapping));
+    }
+
+    /**
+     * Syntactic sugar for {@link #overlaps(Function, Function, Function, Function)} where
+     * both elements of the tuple (A,B) are the same type and have the same interval
+     * defining function. Identical to
+     * {@code overlaps(startMapping, endMapping, startMapping, endMapping)}.
```

Comment:
We should probably use the same trick here as we do for all overloaded methods in the UniConstraintStream API javadocs etc: simply put `As defined by {@link #overlapping(...)}.` and then still specify the @param and @return etc.